src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,390 @@
+ * Once the thread runs the task for the set number of repetitions,
+ * it will pause and wait until the number of remaining repetitions is changed.
+ * <p>
+ * This thread does not finish running until {@link #kill()} or {@link #blockingKill()} is called.
+ * Once started, be sure to kill this thread.
+ * <p>
+ * If this thread is interrupted while paused, a repetition will run such that the interrupted status
+ * can be handled by user code (i.e. code in the passed in task, or overridden {@link #repeat()} method).
+ * <p>
+ * Optionally, you may set a limit to the loop frequency through {@link #setFrequencyLimit(double)},
+ * or by passing the frequency limit as a parameter in the constructor.
+ * The loop frequency limit may be changed at any time. To de-limit the loop frequency,
+ * use {@link #removeFrequencyLimit()}, or set the limit < 0.0.
+ * Setting the frequency limit only guarantees that the loop's frequency will not exceed the limit.
+ * It does not guarantee that the loop will run at the set frequency, as the code executed within the loop
+ * may be too slow to run at that frequency.
+ * <p>
+ * Like {@link Thread}, the {@link RepeatingTaskThread} may be used in two ways:
+ * <ul>
+ *    <li>
+ *       First, by passing in a runnable (or in this case a {@link RunnableThatThrows}) to the constructor.
+ *       This runnable will be called in {@link #repeat()} every repetition.
+ *    <li>
+ *       Second, by {@code @Override}ing the {@link #repeat()} method.
+ *       The code within {@link #repeat()} will run every repetition.
+ */
+public class RepeatingTaskThread extends Thread
+{
+   public static final int REPEAT_INDEFINITELY = -1;
+   public static final double UNLIMITED_FREQUENCY = -1.0;
+
+   private final RunnableThatThrows task;
+   private final ExceptionHandler exceptionHandler;
+   private final Object loopLock = new Object();
+
+   /** Throttler for optionally set loop period/frequency limit */
+   private final Throttler throttler = new Throttler();
+
+   /**
+    * Countdown for number of repetitions to run.
+    * The counter is decremented each time before the repetition.
+    * Once the counter hits 0, the loop is paused.
+    * <ul>
+    *    <li> 0 = pause (don't run the loop until counter value is changed).
+    *    <li> -1 = repeat indefinitely (keep looping until told otherwise).
+    *    <li> N > 0 = run the loop N more repetitions.
+    */
+   private volatile int remainingRepetitions = 0;
+
+   /** Counter for the total number of repetitions completed during the lifetime of this thread. */
+   private long completedRepetitions = 0L;
+
+   /**
+    * Indicates whether this object is destroyed.
+    * The loop will come to a finish when {@code isRunning() == false}.
+    * Does not equate to {@link Thread#isAlive()}, as the thread may take
+    * some time to finish executing after {@code isRunning()} becomes {@code false}.
```

Comment:
I would try and reword this to encapsulate the following, or just replace with the following. I think this variable doesn't have a super meaningful meaning and isn't part of the API so doesn't need a meaningful description.

Initially false. When true, the Thread has been started. When true -> false, the main loop allows the currently executing task (if any) to complete and then allows the thread to die.

Reply:
I changed the comment. Let me know if it's good